aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -90,13 +125,37 @@ struct AT_CUDA_API CUDAEvent {
 
   // Note: cudaStreamWaitEvent must be called on the same device as the stream.
   // The event has no actual GPU resources associated with it.
-  void block(const CUDAStream& stream) {
+  void block(const CUDAStream& stream) const {
     if (is_created_) {
       CUDAGuard guard(static_cast<int16_t>(stream.device_index()));
       AT_CUDA_CHECK(cudaStreamWaitEvent(stream, event_, 0));
     }
   }
 
+  float elapsed_time(const CUDAEvent& other) const {
+    float time_ms = 0;
+    // raise cudaErrorNotReady if either event is recorded but not yet completed
+    AT_CUDA_CHECK(cudaEventElapsedTime(&time_ms, event_, other.event_));
+    return time_ms;
+  }
+
+  void synchronize() const {
+    if (is_created_) {
+      AT_CUDA_CHECK(cudaEventSynchronize(event_));
+    }
+  }
+
+  void ipc_handle(cudaIpcEventHandle_t * handle) {
+    #ifndef __HIP_PLATFORM_HCC__
+      AT_CHECK(is_created_,
+        "Events must be recorded before creating IPC handles.");
+
+      AT_CUDA_CHECK(cudaIpcGetEventHandle(handle, event_));
```

Comment:
I think this also needs to happen on the event's device. From what I've seen it may segfault if called from the wrong device.

(Also, add a test for this please)

Reply:
Will do, thanks!